src/backend/distributed/connection/remote_commands.c

```diff
@@ -652,3 +656,224 @@ FinishConnectionIO(MultiConnection *connection, bool raiseInterrupts)
+			{
+				WaitEvent *event = &events[eventIndex];
+				MultiConnection *connection = NULL;
+				bool connectionIsReady = false;
+
+				if (event->events & WL_POSTMASTER_DEATH)
+				{
+					ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
+				}
+
+				if (event->events & WL_LATCH_SET)
+				{
+					ResetLatch(MyLatch);
+
+					if (raiseInterrupts)
+					{
+						CHECK_FOR_INTERRUPTS();
+					}
+
+					if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
+					{
+						/* return immediately in case of cancellation */
+						FreeWaitEventSet(waitEventSet);
+						return;
+					}
+
+					continue;
+				}
+
+				connection = (MultiConnection *) event->user_data;
+				connectionIndex = event->pos + pendingConnectionsStartIndex;
+
+				if (event->events & WL_SOCKET_WRITEABLE)
+				{
+					int sendStatus = PQflush(connection->pgConn);
+					if (sendStatus == -1)
+					{
+						/* send failed, done with this connection */
+						connectionIsReady = true;
+					}
+					else if (sendStatus == 0)
+					{
+						/* done writing, only wait for read events */
+						ModifyWaitEvent(waitEventSet, connectionIndex, WL_SOCKET_READABLE,
+										NULL);
+					}
+				}
+
+				if (event->events & WL_SOCKET_READABLE)
+				{
+					int receiveStatus = PQconsumeInput(connection->pgConn);
+					if (receiveStatus == 0)
+					{
+						/* receive failed, done with this connection */
+						connectionIsReady = true;
+					}
+					else if (!PQisBusy(connection->pgConn))
+					{
+						/* result was received */
+						connectionIsReady = true;
+					}
+				}
+
+				if (connectionIsReady)
+				{
+					connectionReady[connectionIndex] = true;
+					rebuildWaitEventSet = true;
+				}
+			}
+
+			/* move non-ready connections to the back of the array */
+			for (connectionIndex = pendingConnectionsStartIndex;
+				 connectionIndex < totalConnectionCount; connectionIndex++)
+			{
+				if (connectionReady[connectionIndex])
+				{
+					allConnections[connectionIndex] = allConnections[pendingConnectionsStartIndex];
+					pendingConnectionsStartIndex++;
+				}
+			}
+
+		}
+
+		if (waitEventSet != NULL)
+		{
+			FreeWaitEventSet(waitEventSet);
+		}
+	}
+	PG_CATCH();
+	{
+		/* make sure the epoll file descriptor is always closed */
+		if (waitEventSet != NULL)
+		{
+			FreeWaitEventSet(waitEventSet);
+		}
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+}
+
+
+/*
+ * BuildWaitEventSet creates a WaitEventSet for the given array of connections
+ * which can be used to wait for any of the sockets to become read-ready, or
+ * write-ready in case there is data to send.
+ */
+static WaitEventSet *
+BuildWaitEventSet(MultiConnection **allConnections, int totalConnectionCount,
+				  int pendingConnectionsStartIndex)
+{
+	int pendingConnectionCount = totalConnectionCount - pendingConnectionsStartIndex;
+	WaitEventSet *waitEventSet = NULL;
+	int connectionIndex = 0;
+
+	waitEventSet = CreateWaitEventSet(CurrentMemoryContext, pendingConnectionCount + 2);
```

Comment:
Alternatively just move the WL_POSTMASTER_DEATH / WL_LATCH_SET cases up here, documenting it via that?